Transformation for Allergy

baiju-grh opened the pull request:
Contains transformation for Allergy between Fhir and HealthVault. This is build on top of the Procedure transformation branch.

baiju-grh committed:
Implemented support for Allergy from FHIR to HV and vice-versa

brandonpollett commented on Microsoft.HealthVault.Fhir.UnitTests/ToFhirTests/AddressToFhirTests.cs:
```diff
@@ -0,0 +1,74 @@
+            {
+                Description = "John's Office Address",
+                IsPrimary = true,
+                Street = { "3804 Lakeland ", "Terrace" },
```
Use microsoft's address

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/AllergyIntoleranceToHealthVault.cs:
```diff
@@ -0,0 +1,157 @@
+                allergenType = allergyExtension.GetStringExtension(HealthVaultExtensions.AllergenType);
+                allergy.AllergenCode = allergyExtension.GetExtensionValue<CodeableConcept>(HealthVaultExtensions.AllergenCode)?.ToCodableValue();
+                allergy.Treatment = allergyExtension.GetExtensionValue<CodeableConcept>(HealthVaultExtensions.AllergyTreatement)?.ToCodableValue();
+
```
nit: extra line

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/AllergyIntoleranceToHealthVault.cs:
```diff
@@ -0,0 +1,157 @@
+
+            }
+
+            foreach (var code in allergyIntolerance.Code.Coding)
```
Once we set this once, do we want to continue to looping through additional codings and possibly overwriting it?

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/AllergyIntoleranceToHealthVault.cs:
```diff
@@ -0,0 +1,157 @@
+
+            if (allergyIntolerance.Reaction != null && allergyIntolerance.Reaction.Count > 0)
+            {              
+                foreach (var code in allergyIntolerance.Reaction.FirstOrDefault().Manifestation.FirstOrDefault().Coding)
```
Same here, were you could end up setting reaction multiple times.

brandonpollett replied:
Since Manifestation has a cardinality of 1..* then we should just use first

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/AllergyIntoleranceToHealthVault.cs:
```diff
@@ -0,0 +1,157 @@
+                    new CodedValue(allergyIntolerance.Category.FirstOrDefault().Value.ToString(), FhirCategories.HL7Allergy, HealthVaultVocabularies.Fhir, "")
+                };
+            }
+            else if (!string.IsNullOrWhiteSpace(allergenType))
```
Would it make more sense to optimistically use the allergenType first instead of the category since it will only be set if the the origination came from health vault?

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/AllergyIntoleranceToHealthVault.cs:
```diff
@@ -0,0 +1,157 @@
+
+            allergy.Name.Add(new CodedValue(code, vocabName, family, version));
+        }
+        private static void SetAllergyReaction(this Allergy allergy, string display, string code, string vocabName, string family, string version)
```
nit: add extra line

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/AllergyToFhir.cs:
```diff
@@ -0,0 +1,134 @@
+            allergyInTolerance.Extension.Add(allergyExtension);
+
+            return allergyInTolerance;
+
```
Nit: extra line

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/PractitionerToHealthVault.cs:
```diff
@@ -44,6 +44,12 @@ private static ItemTypes.PersonItem ToPerson(this Hl7.Fhir.Model.Practitioner fh
             {
                 foreach (var address in fhirPractitioner.Address)
                 {
+                    if ((string.IsNullOrEmpty(address.City))||(string.IsNullOrWhiteSpace(address.Line.ToString()))||(string.IsNullOrWhiteSpace(address.PostalCode))
```
If this is a requirement for converting the address to healthvault, this check should be completed in the tohealthvault() method instead of before the call.

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/AllergyToFhir.cs:
```diff
@@ -0,0 +1,134 @@
+            {
+                List<AllergyIntoleranceCategory?> lstAllergyIntoleranceCategory = new List<AllergyIntoleranceCategory?>();
+                AllergyIntoleranceCategory allergyIntoleranceCategory;
+                if (aValue.Value != null)
```
Add extra line

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/AllergyToFhir.cs:
```diff
@@ -0,0 +1,134 @@
+                        allergyExtension.AddExtension(HealthVaultExtensions.AllergenType, new FhirString(aValue.Value));
+                    }
+                }
+                if (lstAllergyIntoleranceCategory.Count > 0)
```
nit: add extra line before

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/AllergyToFhir.cs:
```diff
@@ -0,0 +1,134 @@
+        {
+            foreach (CodedValue aValue in allergenType)
+            {
+                List<AllergyIntoleranceCategory?> lstAllergyIntoleranceCategory = new List<AllergyIntoleranceCategory?>();
```
this will only ever have 1 value in it since it's declared inside the foreach.

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/AllergyToFhir.cs:
```diff
@@ -0,0 +1,134 @@
+            if (allergy.CommonData != null)
+            {
+                var note = new Hl7.Fhir.Model.Annotation();
+                note.Text = allergy.CommonData.Note;
```
Null check `CommonData.Note`

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/AllergyToFhir.cs:
```diff
@@ -0,0 +1,134 @@
+                    }
+                    else
+                    {
+                        allergyExtension.AddExtension(HealthVaultExtensions.AllergenType, new FhirString(aValue.Value));
```
We're possibly adding multiple `allergen-type` extensions into the `allergy` extension. Is this expected?

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/AllergyToFhir.cs:
```diff
@@ -0,0 +1,134 @@
+    }
+    internal static class AllergyToFhir
+    {
+        internal static AllergyIntolerance ToFhirInternal(Allergy allergy, AllergyIntolerance allergyInTolerance)
```
`allergyInTolerance` should be `allergyIntolerance`

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/AllergyToFhir.cs:
```diff
@@ -0,0 +1,134 @@
+            return AllergyToFhir.ToFhirInternal(allergy, ToFhirInternal<AllergyIntolerance>(allergy));
+        }
+    }
+    internal static class AllergyToFhir
```
nit: add extra line before

brandonpollett commented:
The procedure work has been merged into master. Please rebase off master to bring the size of changes down.

brandonpollett reviewed (changes requested):
Please update based on the comments.

baiju-grh committed:
Merge branch 'master' into personal/baiju-grh/allergy

baiju-grh committed:
PR Comments Updated

Merge branch 'master' into personal/baiju-grh/allergy

# Conflicts:
#	Microsoft.HealthVault.Fhir.UnitTests/Microsoft.HealthVault.Fhir.UnitTests.csproj
#	Microsoft.HealthVault.Fhir/Constants/HealthVaultExtensions.cs

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/AllergyIntoleranceToHealthVault.cs:
```diff
@@ -0,0 +1,155 @@
+
+            var coding = allergyIntolerance.Code.Coding.FirstOrDefault();
+
+            if (HealthVaultVocabularies.SystemContainsHealthVaultUrl(coding.System))
```
Coding needs to be null checked before using properties off of it.

baiju-grh replied:
code updated with null check

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/AllergyIntoleranceToHealthVault.cs:
```diff
@@ -0,0 +1,155 @@
+            else
+            {
+                allergy.SetAllergyName(
+                    coding.Display,
```
Coding needs to be null checked before using properties off of it.

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/AllergyIntoleranceToHealthVault.cs:
```diff
@@ -0,0 +1,155 @@
+
+            if (allergyIntolerance.Reaction != null && allergyIntolerance.Reaction.Count > 0)
+            {
+                var code = allergyIntolerance.Reaction.FirstOrDefault().Manifestation.FirstOrDefault().Coding.First();
```
Reaction can just be `First()` since we already checked that it exists.

brandonpollett replied:
If manifestation is guaranteed to be there then we should just use `First()`.

brandonpollett replied:
I believe coding has a cardinality of 0-* usually, so we probably want to do `FirstOrDefault()` and null check it before using.

brandonpollett commented on Microsoft.HealthVault.Fhir/Transformers/AllergyIntoleranceToHealthVault.cs:
```diff
@@ -0,0 +1,155 @@
+            }
+            else if (allergyIntolerance.Category != null && allergyIntolerance.Category.Count() > 0)
+            {
+                allergy.AllergenType = new CodableValue(allergyIntolerance.Category.FirstOrDefault().Value.ToString())
```
Since we've already checked that category exists, let's use `First()`

baiju-grh committed:
PR Comments
